src/state/transactionState.ts

```diff
@@ -0,0 +1,560 @@
+        // Map each group to be sorted by group.index
+        (group) => orderBy(group, [(item) => item?.group?.index], ['asc']),
+      ),
+      // Filter all items that do not contain a transaction of index 0
+      // (this means they were cut off within the group)
+      (group) => !!group.find((item) => item?.group?.index === 0),
+      // Finally we sort the resulting array of arrays by the createdAt value of the first transaction within a group (descending)
+    ).sort(
+      (groupA, groupB) =>
+        new Date(groupB[0]?.createdAt || 0).valueOf() -
+        new Date(groupA[0]?.createdAt || 0).valueOf(),
+    );
+  }, [data?.getTransactionsByUser?.items]);
+
+  const groupState = getGroupStatusAll(transactions, loading);
+
+  return { groupState, transactions, loading };
+};
+
+export const getTransaction = async (id: string, fetchPolicy?: FetchPolicy) => {
+  // Get the transaction from apollo
+  const apolloClient = getContext(ContextModule.ApolloClient);
+
+  const { data } = await apolloClient.query<
+    GetTransactionQuery,
+    GetTransactionQueryVariables
+  >({
+    query: GetTransactionDocument,
+    variables: { id },
+    fetchPolicy,
+  });
+
+  if (!data.getTransaction) {
+    throw new Error('Transaction not found.');
+  }
+  return convertTransactionType(data.getTransaction);
+};
+
+export const fetchTransaction = async ({ id }) => {
+  const apollo = getContext(ContextModule.ApolloClient);
+  return apollo.query<GetTransactionQuery, GetTransactionQueryVariables>({
+    query: GetTransactionDocument,
+    variables: { id },
+  });
+};
+
+export const addTransactionToDb = async (
+  id: string,
+  {
+    context,
+    createdAt,
+    from,
+    group,
+    identifier,
+    methodContext,
+    methodName,
+    options = {},
+    params,
+    status,
+    title,
+    titleValues,
+    metatransaction,
+  }: TransactionCreatedPayload,
+) => {
+  let colonyAddress = '0x';
+  try {
+    // @TODO: ultimately we should get rid of this (once sagas are out)
+    colonyAddress = getContext(ContextModule.CurrentColonyAddress);
+  } catch {
+    // If we don't have a colony address, we're creating a colony.
+    // The correct address will be added to the transactions in the colony create saga
+  }
+
+  const fromAddress = utils.getAddress(from);
+
+  // Apollo needs the null values to be there (for the optimistic response - it doesn't like undefined)
+  const txGroup = {
+    id: `${group.id}-${group.index}`,
+    groupId: group.id.toString(),
+    index: group.index,
+    key: group.key,
+    description: JSON.stringify(group.description) || null,
+    descriptionValues: JSON.stringify(group.descriptionValues) || null,
+    title: JSON.stringify(group.title) || null,
+    titleValues: JSON.stringify(group.titleValues) || null,
+  };
+
+  const txCreatedAt = createdAt.toISOString();
+  const txParams = JSON.stringify(params || []);
+
+  const apollo = getContext(ContextModule.ApolloClient);
+
+  const input = {
```

Comment:
maybe I'd type this just for code completion

Reply:
Gotcha.